### ISSUE TITLE
plist_to_xml function  is very slow

### ISSUE THREAD

**Issue by kxj0o** [libimobiledevice/libplist#81]

**Comment by nikias** [comment]
A bit more information please. Do you have an example file or code? Without any description I don't know where to look...

**Comment by nikias** [comment]
@kxj0o can you give an example please?

**Comment by kxj0o** [comment]
@nikias idevicerestore tool, restore.c file 1606 line call restored_send() send Baseband Data, plist format Baseband Data is very large 29.3 MB

**Comment by kxj0o** [comment]
The original size of the  [BasebandData.txt](https://github.com/libimobiledevice/libplist/files/654227/BasebandData.txt) file is 29.3MB

**Comment by nikias** [comment]
@kxj0o I don't see any issue with that. The plist you attached is 9.1MB and what is the problem with the library sending it? The conversion to xml is actually quite fast.

**Comment by nikias** [comment]
@kxj0o maybe I'm missing the point here, but you haven't really explained/shown why or where it is slow.

To give you some example, the following file is a 162 MB XML plist, that has just one huge <data> node in it:
```
$ ls -l hugedata.plist
-rw-r--r--  1 nikias  staff  169580205 11 Jan 17:14 hugedata.plist
```
Converting it from XML to binary takes about a second:
```
$ time plistutil -i hugedata.plist -o hugedata.bplist

real	0m1.071s
user	0m0.476s
sys	0m0.513s
```
And converting it from binary to XML it takes another second:
```
$ time plistutil -i hugedata.bplist -o hugedata.xml

real	0m0.979s
user	0m0.292s
sys	0m0.551s
```
I don't see anything that would be considered slow here. Also when running the same conversion through Apple's plutil, libplist is actually faster:
```
$ time plutil -convert binary1 -o hugedata.bplist hugedata.plist

real	0m1.343s
user	0m0.702s
sys	0m0.374s
```
and
```
$ time plutil -convert xml1 -o hugedata.xml hugedata.bplist

real	0m1.201s
user	0m0.487s
sys	0m0.341s
```

**Comment by kxj0o** [comment]
@nikias Test data can be downloaded from here [https://1drv.ms/f/s!Agxy6bY8RrCMartAvmorOziqpGc](url)

    plist_t plist = NULL;
    plist_from_file(L"BasebandData.xml", &plist);
    char *plist_xml = NULL;
    uint32_t length = 0;
    uint32_t tick = GetTickCount();
    plist_to_xml(plist, &plist_xml, &length); 
    tick = GetTickCount() - tick;
    
   plist_to_xml()  takes about 45 seconds

**Comment by nikias** [comment]
@kxj0o link is broken

**Comment by kxj0o** [comment]
@nikias paste a link into a browser address bar and open it

**Comment by nikias** [comment]
Ok I just tried on win32 and it is slow. I will investigate. On macOS it takes less than a second to run that code.

**Comment by nikias** [comment]
I figured out what's wrong. The str_buf_append() will realloc a million times and realloc is _REALLY_ slow on windows. I will calculate the size and grow the buffer beforehand so it doesn't reallocate so many times.

**Comment by kxj0o** [comment]
@nikias good idea ^_^

**Comment by xdeng** [comment]
@nikias @kxj0o  OS win7 64   IDE vs2005 debug 21x  release 15x   Not very slow

**Comment by nikias** [comment]
Fixed with commit 8ad21e6b59e5e5a3104660f17ad2b0e9122edecf
